shell/platform/android/android_choreographer.cc

```diff
@@ -0,0 +1,61 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/android/android_choreographer.h"
+
+#include "flutter/fml/native_library.h"
+
+// Only avialalbe on API 24+
+typedef void AChoreographer;
+// Only avialalbe on API 29+ or API 24+ and architecture is 64-bit
+typedef void (*AChoreographer_frameCallback)(int64_t frameTimeNanos,
+                                             void* data);
+// Only avialalbe on API 24+
+typedef AChoreographer* (*AChoreographer_getInstance_FPN)();
+typedef void (*AChoreographer_postFrameCallback_FPN)(
+    AChoreographer* choreographer,
+    AChoreographer_frameCallback callback,
+    void* data);
+static AChoreographer_getInstance_FPN AChoreographer_getInstance;
+static AChoreographer_postFrameCallback_FPN AChoreographer_postFrameCallback;
+
+namespace flutter {
+
+bool AndroidChoreographer::ShouldUseNDKChoreographer() {
+  static std::optional<bool> should_use_ndk_choreographer;
+  if (should_use_ndk_choreographer) {
+    return should_use_ndk_choreographer.value();
+  }
+  auto libandroid = fml::NativeLibrary::Create("libandroid.so");
+  FML_DCHECK(libandroid);
+  auto get_instance_fn =
+      libandroid->ResolveFunction<AChoreographer_getInstance_FPN>(
+          "AChoreographer_getInstance");
+  auto post_frame_callback_fn =
+      libandroid->ResolveFunction<AChoreographer_postFrameCallback_FPN>(
+          "AChoreographer_postFrameCallback64");
+#if FML_ARCH_CPU_64_BITS
+  if (!post_frame_callback_fn) {
+    post_frame_callback_fn =
+        libandroid->ResolveFunction<AChoreographer_postFrameCallback_FPN>(
+            "AChoreographer_postFrameCallback");
+  }
+#endif
+  if (get_instance_fn && post_frame_callback_fn) {
```

Comment:
is this right? 

if `!FML_ARCH_CPU_64_BITS` and `Choreographer_postFrameCallback64 == nullptr`, then  `should_use_ndk_choreographer` returns false.

Reply:
Yes, we can not use `AChoreographer_postFrameCallback` in 32-bits app, because in the callback function `AChoreographer_frameCallback(long frameTimeNanos, void *data)`, the type of `frameTimeNanos` is `long`, and in 32-bits architecture `long` is 32bit. And actually the correct `frameTimeNanos` should be 64 bit. So in 32-bits architecture we can't get the correct `frameTimeNanos`.

Reply:
> I see. I didn't realize this was going to be this complicated.

We can also simplify it, for example , API 29+ use NDK `AChoreographer_postFrameCallback64`, others use Java fallback. So that we don't need to consider whether the app is 32-bits or 64-bits.

Reply:
I think if we are doing this, then making it work on API level >=24 for x64 is better than only >= 29.

Reply:
got it.